### ISSUE TITLE
2.6.4 causes error 500 out of memory on GraphQL queries

### ISSUE THREAD

**Issue by plamenh** [api-platform/core#4220]
**API Platform version(s) affected**: 2.6.4

**Description**  
Upgraded from 2.6.3 to 2.6.4 and started getting out of memory errors (500) on GraphQL queries.

**How to reproduce**  

Upgrade from 2.6.3 to 2.6.4

**Additional Context**  

This has got to be a circular reference issue that started happening with 2.6.4. No matter how much memory I allocate, it does not work. I have narrowed it down to 2 entities with relations:

```
/**
 * @ORM\Table(name="account_alias")
 * @ORM\Entity
 * @ExclusionPolicy("all")
 * @ApiResource(
 *     normalizationContext={"groups"={"read"}, "enable_max_depth"=true},
 *     denormalizationContext={"groups"={"write"}, "enable_max_depth"=true},
 *     attributes={"security"="is_granted('ROLE_USER')", "pagination_enabled"=false}
 * )
 */
class AccountAlias
{
    ...

    /**
     * One AccountAlias has many AccountEvents assigned.
     *
     * @ORM\OneToMany(targetEntity="AccountEvent", mappedBy="assignee")
     * @MaxDepth(1)
     * @SkipWhenEmpty
     * @Groups({"read"})
     * @ApiSubresource
     */
    private $accountEventsAssigned;
    
    /**
     * One AccountAlias has many AccountEvents owned.
     *
     * @ORM\OneToMany(targetEntity="AccountEvent", mappedBy="owner")
     * @Groups({"read"})
     * @ApiSubresource
     */
    private $accountEventsOwned;

    ...
}
```

```
/**
 *
 * @ORM\Table(name="account_event")
 * @ORM\Entity(repositoryClass="App\Repository\AccountEventRepository")
 * @ExclusionPolicy("all")
 * @ApiResource(
 *     attributes={"security"="is_granted('ROLE_USER')", "pagination_enabled"=false}
 * )
 * @ApiFilter(ExistsFilter::class, properties={"createdVia.deletedAt", "accountEventCollection"})
 * @ApiFilter(SearchFilter::class, properties={"subject","assignee","type","owner"})
 * @ApiFilter(DateFilter::class, properties={"dueAt"})
 */
class AccountEvent
{
    ...

    /**
     * Many accounts have one assignee.
     *
     * @ORM\ManyToOne(targetEntity="AccountAlias", inversedBy="accountEventsAssigned")
     * @Expose
     * @MaxDepth(3)
     */
    private $assignee;

    /**
     * Many accounts have one owner.
     *
     * @ORM\ManyToOne(targetEntity="AccountAlias", inversedBy="accountEventsOwned")
     * @Expose
     * @MaxDepth(3)
     * @ApiSubresource
     */
    private $owner;

    ...
}
```

Whenever I try to query accountEventsAssigned on an accountAlias, I get out of memory.

What I am thinking is happening is we have accountEvents with both the owner and assignee being the same accountAlias, causing a circular ref. However, I removed @ApiSubresource from $owner in accountEvent and I still hit the error. The worse part is that it's a 500 error, so I get no trace that I can work with.

Plus, this setup has been working fine up to 2.3.6. It's not clear to me how to work this way (we encountered so many circular reference problems and have been going around them for ages). MaxDepth definitely does not work and this forces us to have only one-way access to data of entities that have multiple relations to the same entity.

**Comment by chuckadams** [comment]
Having the same issue myself as well: 2.6.3 was fine, then boom in 2.6.4.  My entities aren't directly circular, but they are many-to-many, which is naturally cyclic.  I seem to have enough memory to query about 10 items or so, so it's not bombing out right away, but any more than that and I get the OOM error in the log with no backtrace of course (thanks PHP).

**Comment by plamenh** [comment]
Adding to the topic, can someone share a solution for these 'circular' references? This is a quite legit use case and we always have to design around this limitation. I tried MaxDepth on all sides, but it does not seem to work. Maybe I am missing something somewhere.

**Comment by sachingk** [comment]
To solve the circular problem use the following to the class property 

`@ApiProperty(attributes={"fetchEager": false})`

For example 

```
  /**
     * @ORM\ManyToOne(targetEntity=Flow::class, inversedBy="flowSteps")
     * @ORM\JoinColumn(nullable=false)
     * @Groups({"item_query","collection_query", "mutation","read", "write"})
     * @ApiProperty(attributes={"fetchEager": false})
     */
    private $flow;
```

**Comment by plamenh** [comment]
> To solve the circular problem use the following to the class property
> 
> `@ApiProperty(attributes={"fetchEager": false})`
> 
> For example
> 
> ```
>   /**
>      * @ORM\ManyToOne(targetEntity=Flow::class, inversedBy="flowSteps")
>      * @ORM\JoinColumn(nullable=false)
>      * @Groups({"item_query","collection_query", "mutation","read", "write"})
>      * @ApiProperty(attributes={"fetchEager": false})
>      */
>     private $flow;
> ```

This does not work. Still getting out of memory exception with it. Something is wrong here. I've tried everything I could find and still hitting this issue.

**Comment by sachingk** [comment]
> This does not work. Still getting out of memory exception with it. Something is wrong here. I've tried everything I could find and still hitting this issue. Could the issue be I am trying this on an ApiSubresource?

Are you  sure you are using GraphQL in your example ? I didn't see any "graphql" configurations in the annotations. 

On the other hand, I above code is a working sample from my project. It's working for me with GraphQL requests.

**Comment by plamenh** [comment]
@sachingk I am not aware of any graphql specific configurations, but I can assure you I am using graphql.

**Comment by plamenh** [comment]
So no idea what changed in 2.6.4?

**Comment by alanpoulain** [comment]
Maybe it's because of this PR: https://github.com/api-platform/core/pull/3525.
You can see all the changes here: https://github.com/api-platform/core/compare/v2.6.3...v2.6.4.

**Comment by plamenh** [comment]
@alanpoulain thanks for the tip. Found the issue. So, for anyone having out of memory exceptions with ManyToMany, 2.6.4 and higher, make sure you set **force_eager** to **false**.

This did it for me. I don't know why it was set to true (I copied the config file at initial set-up and it came like that) and whether it is supposed to be so. The fact of the matter is this does not work with ManyToMany relations.

Thanks again! This was a MAJOR headache.
